Applications/Spire/Include/Spire/Ui/ContextMenu.hpp

```diff
@@ -0,0 +1,106 @@
+#ifndef SPIRE_CONTEXT_MENU_HPP
+#define SPIRE_CONTEXT_MENU_HPP
+#include <QWidget>
+#include "Spire/Spire/ListModel.hpp"
+#include "Spire/Ui/CheckBox.hpp"
+#include "Spire/Ui/Ui.hpp"
+
+namespace Spire {
+
+  /** Represents a pop-up context menu. */
+  class ContextMenu : public QWidget {
+    public:
+
+      /**
+       * Signals that a menu item was submitted.
+       * @param menu The ContextMenu where the submission is triggered.
+       * @param label The label of the submitted the menu item.
+       */
+      using SubmitSignal = Signal<void (const ContextMenu& menu,
+        const QString& label)>;
+
+      /**
+       * The type of function used to do an action.
+       * @param name The name of the action.
+       */
+      using Action = std::function<void (const QString& name)>;
+
+      /**
+       * Constructs a ContextMenu.
+       * @param parent The parent widget.
+       */
+      explicit ContextMenu(QWidget& parent);
+
+      /**
+       * Adds a menu.
+       * @param name The name of the menu.
+       * @param menu The menu to add.
+       */
+      void add_menu(const QString& name, const ContextMenu& menu);
+
+      /**
+       * Adds an action.
+       * @param name The name of the action.
+       * @param action The action called when the item is submitted.
+       */
+      void add_action(const QString& name, const Action& action);
+
+      /**
+       * Adds a CheckBox.
+       * @param name The name of the menu item.
+       * @return The checked state.
+       */
+      std::shared_ptr<BooleanModel> add_check_box(const QString& name);
+
+      /**
+       * Adds a CheckBox.
+       * @param name The name of the menu item.
+       * @param checked The checked state.
+       */
+      void add_check_box(const QString& name,
+        const std::shared_ptr<BooleanModel>& checked);
+
+      /** Connects a slot to the Submit signal. */
+      boost::signals2::connection connect_submit_signal(
+        const SubmitSignal::slot_type& slot) const;
+
+    protected:
+      bool eventFilter(QObject* watched, QEvent* event) override;
+      bool event(QEvent* event) override;
+
+    private:
+      enum class MenuItemType {
+        ACTION,
+        CHECK,
+        SUBMENU
+      };
+      struct MenuItem {
+        MenuItemType m_type;
+        QString m_name;
+        std::any m_data;
```

Comment:
Why is a ```std::any``` used?

Reply:
`m_data` could be a `ContextMenu` component for a Submenu, an `Action` function for a action and a `BooleanModel` for the check box, so I used `std::any` to hold them.

Reply:
```std::any``` is used to represent an open set of heterogeneous types. ```std::variant``` is used to represent a closed set of heterogeneous types.